### PR TITLE
feat(webapp): safe dataset formats — .npz and Parquet convert at upload

### PR DESCRIPTION
**Pull request by fazelehh**
Implements the safe-format recommendation from #453. Stacked on #455 (merge that first; this then retargets).

**The gap this closes:** #455's authentication decides *who* may call the server — it does nothing about *what* a file does when loaded. Auditing a model you did not train (model zoo, Kaggle, a vendor checkpoint) is a core LeakPro use case, and a hostile pickled dataset executes on the operator's machine the moment it is inspected.

**How:** new `dataio.py` converts safe formats once, at the trust boundary:

- `.npz` — `data` + `targets` entries, loaded with `allow_pickle=False`
- Parquet / CSV / JSON(L) — column names become `feature_names`, label column auto-detected
- bare `.npy` — data array, `allow_pickle=False`

The upload becomes a **server-generated** `ArrayDataset` pickle, so everything downstream (training path, audit worker, core `_load_population`) keeps consuming a pickle — but of a known class, written by the server. The uploaded bytes are never unpickled by anyone. A safe-format file smuggling pickled objects (object arrays) is rejected with 400, never handed to a pickle fallback.

`ArrayDataset` keeps raw arrays in `.data`/`.targets` (the interface `sample_data`/`sample_image` and core's population indexing already rely on) and normalises in `__getitem__` exactly as the training path did.

**Legacy `.pkl` datasets still work** for files the operator created; the upload UI now warns when a pickle-based file is selected and recommends `.npz`/Parquet for third-party data. `SECURITY.md` replaces the "treat every uploaded file as trusted input" guidance with the honest per-format reality.

**Verification:** 7 new tests (`test_webapp_dataio.py`) — npz and Parquet round-trips through the endpoint and `sample_data`, hostile npz rejected at upload, legacy fallback intact; all 18 webapp tests pass; `tsc` and `vite build` clean.

Refs #453.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: leakpro/webapp/backend/dataio.py [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+        with np.load(path, allow_pickle=False) as npz:
+            data = _pick(npz, _DATA_KEYS)
+            targets = _pick(npz, _TARGET_KEYS)
+            names = list(npz.files)
+    except ValueError as e:
+        # Raised for object (pickled) arrays — refuse rather than fall back.
+        raise HTTPException(
+            status_code=400,
+            detail="The .npz contains pickled objects; only plain arrays are accepted.",
+        ) from e
```
Comment (Muhaddisabarat) [inline leakpro/webapp/backend/dataio.py:98]: **Only `ValueError` is caught — a corrupt/truncated `.npz` raises `zipfile.BadZipFile`, which isn't a `ValueError`.**

Reproduced: truncating a valid `.npz` mid-file (breaking the zip central directory) raises `zipfile.BadZipFile: File is not a zip file` from `np.load()`. Since `upload_data()` has no surrounding try/except, this becomes an unhandled 500 on the main upload endpoint instead of the 400 this module promises.

```python
# add near the top of the file: import zipfile

def _from_npz(path: Path) -> tuple[np.ndarray, np.ndarray, None]:
    try:
        with np.load(path, allow_pickle=False) as npz:
            data = _pick(npz, _DATA_KEYS)
            targets = _pick(npz, _TARGET_KEYS)
            names = list(npz.files)
    except (ValueError, zipfile.BadZipFile, OSError, EOFError) as e:
        # ValueError: pickled/object arrays. BadZipFile/OSError/EOFError: a
        # truncated or otherwise corrupt upload, not necessarily malicious --
        # either way this boundary must reject with 400, never raise raw.
        raise HTTPException(
            status_code=400,
            detail="Could not read the .npz file (corrupt, or contains pickled objects).",
        ) from e
```

File: leakpro/webapp/backend/dataio.py [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+def _from_npy(path: Path) -> tuple[np.ndarray, np.ndarray, None]:
+    try:
+        data = np.load(path, allow_pickle=False)
+    except ValueError as e:
+        raise HTTPException(
+            status_code=400,
+            detail="The .npy contains pickled objects; only plain arrays are accepted "
+                   "(or use .npz with data + targets entries).",
+        ) from e
+    return data, np.zeros(len(data), dtype=np.int64), None
```
Comment (Muhaddisabarat) [inline leakpro/webapp/backend/dataio.py:131]: **Same gap here — an empty `.npy` raises `EOFError`, not `ValueError`.**

Reproduced: `np.load()` on a zero-byte `.npy` raises `EOFError: No data left in file`, uncaught by this except clause.

```python
def _from_npy(path: Path) -> tuple[np.ndarray, np.ndarray, None]:
    try:
        data = np.load(path, allow_pickle=False)
    except (ValueError, OSError, EOFError) as e:
        raise HTTPException(
            status_code=400,
            detail="Could not read the .npy file (corrupt, empty, or contains pickled objects) "
                   "(or use .npz with data + targets entries).",
        ) from e
    return data, np.zeros(len(data), dtype=np.int64), None
```

File: leakpro/webapp/backend/dataio.py [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+        df = pd.read_parquet(path)
+    elif suffix == ".csv":
+        df = pd.read_csv(path)
+    else:
+        df = pd.read_json(path, lines=(suffix == ".jsonl"))
+
+    label_col = next((c for c in _LABEL_COLUMNS if c in df.columns), None)
+    if label_col is None:
+        label_col = df.columns[-1]
+    features = df.drop(columns=[label_col])
```
Comment (Muhaddisabarat) [inline leakpro/webapp/backend/dataio.py:153]: **`pd.read_parquet`/`pd.read_csv`/`pd.read_json` (and the `df.columns[-1]` fallback) aren't guarded at all** — only the later `.to_numpy()` conversion is (lines 138-145). A malformed Parquet/CSV/JSON file, or one with zero columns, raises whatever pandas/pyarrow/the JSON decoder raises natively and propagates as an unhandled 500 via `upload_data()`.

```python
def _from_table(path: Path) -> tuple[np.ndarray, np.ndarray, list]:
    import pandas as pd
    suffix = path.suffix.lower()
    try:
        if suffix == ".parquet":
            df = pd.read_parquet(path)
        elif suffix == ".csv":
            df = pd.read_csv(path)
        else:
            df = pd.read_json(path, lines=(suffix == ".jsonl"))
    except Exception as e:
        raise HTTPException(status_code=400, detail=f"Could not parse {suffix} file: {e}") from e

    if df.shape[1] == 0:
        raise HTTPException(status_code=400, detail="File has no columns.")
    label_col = next((c for c in _LABEL_COLUMNS if c in df.columns), None)
    if label_col is None:
        label_col = df.columns[-1]
    features = df.drop(columns=[label_col])
```

**Review by Muhaddisabarat — COMMENTED** [review]
Reviewed `dataio.py`, the `main.py`/`security.py` wiring, the frontend warning UI, and `SECURITY.md`.

The core design is sound and I verified the central claim holds: a `.npz` with a pickled/object array (`test_npz_with_pickled_objects_rejected`) is correctly caught in `_from_npz` and turned into a 400 — it never reaches a pickle loader. `set_data_path` also correctly re-raises `HTTPException` before its generic catch-all, so `convert_upload`'s specific error messages survive there. The `ArrayDataset` normalisation, the frontend pickle warning, and the `SECURITY.md` rewrite all look correct.

One class of issue found, in three places — see the inline comments on `dataio.py`. Summary: the exception handling in `_from_npz`/`_from_npy`/`_from_table` only anticipates the *adversarial* failure mode (pickled objects), not a merely *corrupt* upload (truncated file, empty file, malformed table). I reproduced both gaps locally:

```
truncated .npz  -> zipfile.BadZipFile  (not a ValueError)
empty .npy      -> EOFError            (not a ValueError)
```

Because `upload_data()` in `main.py` has no try/except around `convert_upload()` (unlike `set_data_path`, which has a catch-all), any of these produce an unhandled 500 on the primary upload flow instead of the clean 400 this module's own docstring promises ("Malformed safe-format files raise `HTTPException(400)` ... they are never handed to a pickle-based fallback"). None of the 7 new tests exercise a structurally corrupt file, only well-formed containers with "wrong" semantic content, so this wasn't caught. Not a security hole on its own (no info leak, still behind auth), but it breaks the guarantee this PR is specifically about, on the more likely real-world case (a bad/partial upload) rather than the adversarial one.

**Comment by fazelehh** [comment]
Thanks @Muhaddisabarat — agreed on all three, and the framing is right: the module promised 400 on malformed input but only delivered it for the adversarial case, not the far more likely corrupt/truncated upload. Fixed in b4939f3e, essentially as you suggested:

- `_from_npz` catches `(ValueError, zipfile.BadZipFile, OSError, EOFError)`
- `_from_npy` catches `(ValueError, OSError, EOFError)`
- `_from_table` guards the `read_parquet`/`read_csv`/`read_json` call itself and rejects zero-column frames

Reproduced each failure before fixing (truncated `.npz` → `BadZipFile`, empty `.npy` → `EOFError`, garbage parquet → pyarrow error), and added the missing test class: structurally corrupt files at the unit level plus a truncated `.npz` through the upload endpoint — the 500-path that none of the original 7 tests exercised. 26/26 webapp tests pass.

Also note: with #455 merged, this PR is retargeted to `main`.

File: leakpro/webapp/backend/dataio.py [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+def _from_npy(path: Path) -> tuple[np.ndarray, np.ndarray, None]:
+    try:
+        data = np.load(path, allow_pickle=False)
+    except (ValueError, OSError, EOFError) as e:
+        raise HTTPException(
+            status_code=400,
+            detail="Could not read the .npy file (corrupt, empty, or contains pickled "
+                   "objects); or use .npz with data + targets entries.",
+        ) from e
+    return data, np.zeros(len(data), dtype=np.int64), None
```
Comment (Muhaddisabarat) [inline leakpro/webapp/backend/dataio.py:131]: A scalar (0-dimensional) `.npy` array slips past the `except (ValueError, OSError, EOFError)` above and crashes here with an uncaught `TypeError`, not a clean 400.

Reproduced:
```python
import numpy as np
np.save('x.npy', np.array(5))
d = np.load('x.npy', allow_pickle=False)  # loads fine, ndim == 0
len(d)  # TypeError: len() of unsized object
```
`np.load()` succeeds (it's a structurally valid file), so the `try/except` around it never fires — the `TypeError` comes from `np.zeros(len(data), ...)` on the next line.

Same root cause applies in `_from_npz` above (line 109's `len(targets) != len(data)`) if either array is 0-d.

Suggest rejecting 0-d arrays explicitly before the `len()` calls, e.g.:
```python
if data.ndim == 0:
    raise HTTPException(status_code=400, detail="Expected an array, got a scalar .npy file.")
```
(and similarly for `targets` in `_from_npz`).

File: leakpro/webapp/backend/main.py [hunk 5]
```diff
@@ -473,7 +474,14 @@ async def upload_data(job_id: str, file: UploadFile) -> DataMeta:
     job = _get_job(job_id)
     dest = _job_dir(job_id) / f"data{safe_suffix(file.filename)}"
     save_upload(file.file, dest)
-    meta = inspect(dest)
+    # Safe formats (.npz/.parquet/.csv/...) are converted here, at the trust
+    # boundary, into a server-generated pickle — the uploaded bytes are never
+    # unpickled. Legacy pickle formats fall through to inspect().
+    converted = convert_upload(dest, _job_dir(job_id))
```
Comment (Muhaddisabarat) [inline leakpro/webapp/backend/main.py]: Unlike `set_data_path` below (which this PR just wrapped in `try: ... except HTTPException: raise / except Exception: -> 400`), `upload_data` has no try/except around `convert_upload` at all. Any non-`HTTPException` failure inside `dataio.py` (e.g. the 0-d array `TypeError` noted on `dataio.py:124`, or anything similar in the future) becomes a raw, unhandled 500 on the primary upload endpoint — the same "promised 400, delivered 500" gap this PR's last commit fixed elsewhere, just via a different trigger.

Suggest mirroring `set_data_path`'s guard here:
```python
try:
    converted = convert_upload(dest, _job_dir(job_id))
except HTTPException:
    raise
except Exception as e:
    raise HTTPException(status_code=400, detail=safe_detail(e, "Failed to convert upload")) from e
```

**Comment by fazelehh** [comment]
Both addressed in 2b52a349 — and the scalar catch is a good one: `np.load` succeeds on a 0-d array so the corrupt-file except never fires, and the `TypeError` came from the `len()` a line later. Reproduced before fixing.

- `_from_npy` and `_from_npz` now reject 0-dimensional arrays explicitly with 400 (both `data` and `targets` in the npz case, covering the `len(targets) != len(data)` line you pointed at).
- `upload_data` now mirrors `set_data_path`'s guard: `HTTPException` passes through, anything unforeseen is logged server-side and returned as a 400 via `safe_detail` — so the next `dataio` edge case degrades to a clean 400 instead of a 500.

Tests: scalar `.npy` and 0-d `.npz` at the unit level, plus a monkeypatched converter raising `RuntimeError` through the endpoint, asserting 400 and that the internal message doesn't leak into the response. 28/28 webapp tests pass.

**Review by Muhaddisabarat — APPROVED** [review]
All 5 outstanding review comments are resolved:
- The 3 error-handling gaps (`BadZipFile`, `EOFError`, unguarded table parsing) fixed in `b4939f3e`.
- The 0-d scalar array crash and the unguarded `convert_upload` call in `upload_data` fixed in `2b52a349`, both with new regression tests (`test_scalar_arrays_rejected`, `test_unexpected_converter_error_becomes_400`).

CI is green. Looks good to merge.
